Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/locate.h

```diff
@@ -0,0 +1,1886 @@
+///
+/// \tparam TriangleMesh A model of `FaceListGraph`
+/// \tparam AABBTraits A model of `AABBTraits` used to define a \cgal `AABB_tree`.
+/// \tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+///
+/// \param tm a triangulated surface mesh
+/// \param outTree output parameter that stores the computed `AABB_tree`
+/// \param np an optional sequence of \ref pmp_namedparameters "Named Parameters" among the ones listed below:
+///
+/// \cgalNamedParamsBegin
+///   \cgalParamBegin{vertex_point_map}
+///     the property map with the points associated to the vertices of `tm`.
+///     If this parameter is omitted, an internal property map for
+///     `boost::vertex_point_t` should be available in `TriangleMesh`.
+///   \cgalParamEnd
+/// \cgalNamedParamsEnd
+///
+template <typename TriangleMesh, typename AABBTraits, typename NamedParameters>
+void build_AABB_tree(const TriangleMesh& tm,
+                     AABB_tree<AABBTraits>& outTree,
+                     const NamedParameters& np)
+{
+  typedef typename GetVertexPointMap<TriangleMesh, NamedParameters>::const_type     VertexPointMap;
+
+  const VertexPointMap vpm = boost::choose_param(boost::get_param(np, internal_np::vertex_point),
+                                                 get_const_property_map(boost::vertex_point, tm));
+
+  return internal::build_AABB_tree(tm, outTree, vpm);
+}
+
+template <typename TriangleMesh, typename AABBTraits>
+void build_AABB_tree(const TriangleMesh& tm,
+                     AABB_tree<AABBTraits>& outTree)
+{
+  return build_AABB_tree(tm, outTree, parameters::all_default());
+}
+
+/// \brief Returns the face location nearest to the given point, as a `Face_location`.
+///
+/// \tparam TriangleMesh A model of `FaceListGraph`
+/// \tparam AABBTraits A model of `AABBTraits` used to define a \cgal `AABB_tree`.
+/// \tparam NamedParameters a sequence of \ref pmp_namedparameters "Named Parameters"
+///
+/// \param p the point to locate on the input triangulated surface mesh
+/// \param tree An `AABB_tree` containing the triangular faces of the input surface mesh to perform the point location with
+/// \param tm a triangulated surface mesh
+/// \param np an optional sequence of \ref pmp_namedparameters "Named Parameters" among the ones listed below:
+///
+/// \cgalNamedParamsBegin
+///   \cgalParamBegin{vertex_point_map}
+///     the property map with the points associated to the vertices of `tm`.
+///     If this parameter is omitted, an internal property map for
+///     `boost::vertex_point_t` should be available in `TriangleMesh`.
+///   \cgalParamEnd
+/// \cgalNamedParamsEnd
+///
+template <typename TriangleMesh, typename AABBTraits, typename NamedParameters>
+typename internal::Locate_types<TriangleMesh>::Face_location
+locate_with_AABB_tree(const typename internal::Locate_types<TriangleMesh, NamedParameters>::Point& p,
+                      const AABB_tree<AABBTraits>& tree,
+                      const 
```

Comment:
I'm a bit surprised it is a ray and not a line.

Reply:
The line does not define a closest point? (Or I missed something in your comment)

Reply:
well the ray might be on the wrong side of the closest face and you'll miss it (or I missed that the direction always points toward the face).

Reply:
I missed the fact that the ray was the input so it's not your pb.